### ISSUE TITLE
mypy infers conditional expression of Dict[str, object] to be object instead of union

### ISSUE THREAD

**Issue by scascketta** [python/mypy#7835]
* Are you reporting a bug, or opening a feature request?

Bug
* Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. 
```python3
from typing import Dict
from dataclasses import dataclass


@dataclass
class Foo:
    bar: bool


@dataclass
class Bar:
    baz: bool


foo_registry: Dict[str, Foo] = {}
bar_registry: Dict[str, Bar] = {}

use_foo = True

registry = foo_registry if use_foo else bar_registry
registry.get('asdfsdf')
```
* What is the actual behavior/output?

On the last line, mypy infers that `registry` has the type `object`
```
error: "object" has no attribute "get"
```

* What is the behavior/output you expect?

mypy infers that the type of `registry` to be `Union[Dict[str, Foo], Dict[str, Bar]]` but it looks like this behavior may have been undone by https://github.com/python/mypy/pull/5095 ?

* What are the versions of mypy and Python you are using?

mypy 0.740 and Python 3.7.2

Do you see the same issue after installing mypy from Git master?

I ran into other issues doing this, but I am on the latest release.

* What are the mypy flags you are using? (For example --strict-optional)

I'm using `--strict --warn-unreachable` though this occurs with no flags at all.

Thanks for reading.

**Comment by JukkaL** [comment]
Yeah, it would be better if the inferred type for `registry` would be a union. However, even `Mapping[str, object]` would be significantly better than `object`.

**Comment by TH3CHARLie** [comment]
Since I worked on #7780 before, I believe the problem is located in the code snippet below
https://github.com/python/mypy/blob/3b5a62e43da5758cb2c3fff40af5b9a368ea698e/mypy/checkexpr.py#L3567-L3571
There's no explicit type annotation to declare `registry` as a Union, so mypy takes the else branch here and simply join the two types from cond expr into an `object` type
A simple workaround for @scascketta 's case would be:
```python
from typing import Dict
from dataclasses import dataclass


@dataclass
class Foo:
    bar: bool


@dataclass
class Bar:
    baz: bool


foo_registry: Dict[str, Foo] = {}
bar_registry: Dict[str, Bar] = {}

use_foo = True
# annotate registry
registry: Union[Dict[str, Foo], Dict[str, Bar]] = foo_registry if use_foo else bar_registry
registry.get('asdfsdf')
```
Interesting enough that the `TODO` in the code has questioned about more Union types, so maybe we should look into that.

**Comment by crusaderky** [comment]
I believe I stumbled on this while creating a list of heterogeneous dicts:
```python
a = {1: {1: 2}}
b = {1: {1: 3}}
c = {2: {1: "x"}}
d = [a, b]
e = [a, c]
reveal_type(a)  # builtins.dict[builtins.int*, builtins.dict*[builtins.int*, builtins.int*]]
reveal_type(b)  # builtins.dict[builtins.int*, builtins.dict*[builtins.int*, builtins.int*]]
reveal_type(c)  # builtins.dict[builtins.int*, builtins.dict*[builtins.int*, builtins.str*]]
reveal_type(d)  # builtins.list[builtins.dict*[builtins.int, builtins.dict[builtins.int, builtins.int]]]
reveal_type(e)  # builtins.list[builtins.object*]
```
Could you confirm it's the same issue? e.g. is the type of ``[a, c]`` inferred using the same algorithm as the type as ``a if a else c``?

It's also interesting to note that ``a or c`` works fine.

**Comment by hauntsaninja** [comment]
Fixed by #17427
